File: board.py
class Board:
    def __init__(self):
        # 64-square array for the board.
        # Ranks 1-8 are indices 56-63 down to 0-7.
        # Let's map index = rank * 8 + file, where a1 is 56, h8 is 7?
        # Standard convention: a1 = 56, b1 = 57, ..., h8 = 7 (so a8=0...h8=7).
        # Actually, let's use: index 0 is a8, 63 is h1.
        self.squares = ['.'] * 64
        self.side_to_move = 'w' # 'w' or 'b'
        self.castling_rights = {'K': False, 'Q': False, 'k': False, 'q': False}
        self.en_passant = None # index of en passant target square
        self.half_move_clock = 0
        self.full_move_number = 1
        self.history = [] # Stack to keep track of state for unmake_move
        
        # Parse initial FEN
        self.parse_fen("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1")

    def parse_fen(self, fen):
        parts = fen.split(' ')
        
        # 1. Piece placement
        rank = 0
        file = 0
        for char in parts[0]:
            if char == '/':
                rank += 1
                file = 0
            elif char.isdigit():
                for _ in range(int(char)):
                    self.squares[rank * 8 + file] = '.'
                    file += 1
            else:
                self.squares[rank * 8 + file] = char
                file += 1
                
        # 2. Side to move
        self.side_to_move = parts[1]
        
        # 3. Castling rights
        self.castling_rights = {'K': False, 'Q': False, 'k': False, 'q': False}
        if parts[2] != '-':
            for char in parts[2]:
                if char in self.castling_rights:
                    self.castling_rights[char] = True
                    
        # 4. En passant
        if parts[3] != '-':
            file_idx = ord(parts[3][0]) - ord('a')
            rank_idx = 8 - int(parts[3][1])
            self.en_passant = rank_idx * 8 + file_idx
        else:
            self.en_passant = None
            
        # 5. Halfmove clock
        self.half_move_clock = int(parts[4]) if len(parts) > 4 else 0
        
        # 6. Fullmove number
        self.full_move_number = int(parts[5]) if len(parts) > 5 else 1

    def print_board(self):
        print("\n  a b c d e f g h")
        print("  ---------------")
        for rank in range(8):
            row_str = str(8 - rank) + "|"
            for file in range(8):
                piece = self.squares[rank * 8 + file]
                row_str += piece + " "
            print(row_str + "|" + str(8 - rank))
        print("  ---------------")
        print("  a b c d e f g h\n")
        print(f"Side to move: {'White' if self.side_to_move == 'w' else 'Black'}")
        rights = "".join([k for k, v in self.castling_rights.items() if v])
        print(f"Castling rights: {rights if rights else '-'}")
        if self.en_passant is not None:
            ep_file = chr((self.en_passant % 8) + ord('a'))
            ep_rank = str(8 - (self.en_passant // 8))
            print(f"En passant: {ep_file}{ep_rank}")
        else:
            print("En passant: -")
            
    def get_piece(self, sq):
        return self.squares[sq]
        
    def set_piece(self, sq, piece):
        self.squares[sq] = piece

    def make_move(self, move):
        # Save state
        state = {
            'castling_rights': self.castling_rights.copy(),
            'en_passant': self.en_passant,
            'half_move_clock': self.half_move_clock
        }
        self.history.append(state)
        
        piece_moved = self.squares[move.start_sq]
        
        # Move piece
        self.squares[move.target_sq] = piece_moved
        self.squares[move.start_sq] = '.'
        
        # Handle en passant capture
        if move.is_en_passant:
            # The captured pawn is on the same rank as the start_sq and same file as target_sq
            ep_capture_sq = move.target_sq + (8 if self.side_to_move == 'w' else -8)
            self.squares[ep_capture_sq] = '.'
            
        # Handle promotion
        if move.promotion_piece != '.':
            self.squares[move.target_sq] = move.promotion_piece
            
        # Handle castling (move the rook)
        if move.is_castling:
            if move.target_sq - move.start_sq == 2: # Kingside
                self.squares[move.target_sq - 1] = self.squares[move.target_sq + 1]
                self.squares[move.target_sq + 1] = '.'
            elif move.start_sq - move.target_sq == 2: # Queenside
                self.squares[move.target_sq + 1] = self.squares[move.target_sq - 2]
                self.squares[move.target_sq - 2] = '.'
                
        # Update en_passant square
        self.en_passant = None
        if piece_moved.lower() == 'p' and abs(move.target_sq - move.start_sq) == 16:
            self.en_passant = move.start_sq + (8 if self.side_to_move == 'b' else -8)
            
        # Update castling rights
        # King moves
        if piece_moved == 'K':
            self.castling_rights['K'] = False
            self.castling_rights['Q'] = False
        elif piece_moved == 'k':
            self.castling_rights['k'] = False
            self.castling_rights['q'] = False
            
        # Rook moves or captures
        rooks = {63: 'K', 56: 'Q', 7: 'k', 0: 'q'}
        for sq in [move.start_sq, move.target_sq]:
            if sq in rooks:
                self.castling_rights[rooks[sq]] = False
                
        # Update half move clock
        if piece_moved.lower() == 'p' or move.piece_captured != '.':
            self.half_move_clock = 0
        else:
            self.half_move_clock += 1
            
        # Update turn and full move
        if self.side_to_move == 'b':
            self.full_move_number += 1
        self.side_to_move = 'b' if self.side_to_move == 'w' else 'w'
        
    def unmake_move(self, move):
        self.side_to_move = 'w' if self.side_to_move == 'b' else 'b'
        if self.side_to_move == 'b':
            self.full_move_number -= 1
            
        # Restore state
        state = self.history.pop()
        self.castling_rights = state['castling_rights']
        self.en_passant = state['en_passant']
        self.half_move_clock = state['half_move_clock']
        
        piece_moved = move.promotion_piece if move.promotion_piece != '.' else self.squares[move.target_sq]
        
        # Undo promotion
        if move.promotion_piece != '.':
            self.squares[move.target_sq] = 'P' if self.side_to_move == 'w' else 'p'
            
        # Move piece back
        self.squares[move.start_sq] = self.squares[move.target_sq]
        self.squares[move.target_sq] = move.piece_captured
        
        # Restore en passant capture
        if move.is_en_passant:
            ep_capture_sq = move.target_sq + (8 if self.side_to_move == 'w' else -8)
            self.squares[ep_capture_sq] = 'p' if self.side_to_move == 'w' else 'P'
            self.squares[move.target_sq] = '.'
            
        # Restore castling (move the rook back)
        if move.is_castling:
            if move.target_sq - move.start_sq == 2: # Kingside
                self.squares[move.target_sq + 1] = self.squares[move.target_sq - 1]
                self.squares[move.target_sq - 1] = '.'
            elif move.start_sq - move.target_sq == 2: # Queenside
                self.squares[move.target_sq - 2] = self.squares[move.target_sq + 1]
                self.squares[move.target_sq + 1] = '.'

File: src/engine/board.js
// ── Board representation ──────────────────────────────────────────────────
export const START_FEN = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1';
// A move from or to a rook's home corner removes that castling right.
const ROOK_CORNERS = { 63: 'K', 56: 'Q', 7: 'k', 0: 'q' };

export class Board {
  constructor() {
    this.squares = Array(64).fill('.');
    this.sideToMove = 'w';
    this.castlingRights = { K: false, Q: false, k: false, q: false };
    this.enPassant = null;
    this.halfMoveClock = 0;
    this.fullMoveNumber = 1;
    this.history = [];
    this.parseFen(START_FEN);
  }

  parseFen(fen) {
    const parts = fen.split(' ');
    this.squares = Array(64).fill('.');
    let rank = 0, file = 0;
    for (const ch of parts[0]) {
      if (ch === '/') { rank++; file = 0; }
      else if (ch >= '1' && ch <= '8') { file += parseInt(ch); }
      else { this.squares[rank * 8 + file] = ch; file++; }
    }
    this.sideToMove = parts[1] || 'w';
    this.castlingRights = { K: false, Q: false, k: false, q: false };
    if (parts[2] && parts[2] !== '-') {
      for (const ch of parts[2]) if (ch in this.castlingRights) this.castlingRights[ch] = true;
    }
    if (parts[3] && parts[3] !== '-') {
      const f = parts[3].charCodeAt(0) - 97;
      const r = 8 - parseInt(parts[3][1]);
      this.enPassant = r * 8 + f;
    } else {
      this.enPassant = null;
    }
    this.halfMoveClock = parts[4] ? parseInt(parts[4]) : 0;
    this.fullMoveNumber = parts[5] ? parseInt(parts[5]) : 1;
    this.history = [];
  }

  reset() { this.parseFen(START_FEN); }

  toFen() {
    const ranks = [];
    for (let r = 0; r < 8; r++) {
      let row = '', empty = 0;
      for (let f = 0; f < 8; f++) {
        const p = this.squares[r * 8 + f];
        if (p === '.') { empty++; continue; }
        if (empty) { row += empty; empty = 0; }
        row += p;
      }
      ranks.push(empty ? row + empty : row);
    }
    const rights = ['K', 'Q', 'k', 'q'].filter(c => this.castlingRights[c]).join('') || '-';
    const ep = this.enPassant === null ? '-'
      : `${String.fromCharCode(97 + (this.enPassant % 8))}${8 - Math.floor(this.enPassant / 8)}`;
    return `${ranks.join('/')} ${this.sideToMove} ${rights} ${ep} ${this.halfMoveClock} ${this.fullMoveNumber}`;
  }

  serialize() {
    return {
      squares: [...this.squares],
      sideToMove: this.sideToMove,
      castlingRights: { ...this.castlingRights },
      enPassant: this.enPassant,
      halfMoveClock: this.halfMoveClock,
      fullMoveNumber: this.fullMoveNumber,
    };
  }

  loadFrom(state) {
    this.squares = [...state.squares];
    this.sideToMove = state.sideToMove;
    this.castlingRights = { ...state.castlingRights };
    this.enPassant = state.enPassant;
    this.halfMoveClock = state.halfMoveClock;
    this.fullMoveNumber = state.fullMoveNumber;
    this.history = [];
  }

  makeMove(move) {
    this.history.push({
      castlingRights: { ...this.castlingRights },
      enPassant: this.enPassant,
      halfMoveClock: this.halfMoveClock,
    });

    const pieceMoved = this.squares[move.startSq];
    this.squares[move.targetSq] = pieceMoved;
    this.squares[move.startSq] = '.';

    if (move.isEnPassant) {
      const capSq = move.targetSq + (this.sideToMove === 'w' ? 8 : -8);
      this.squares[capSq] = '.';
    }
    if (move.promotionPiece !== '.') {
      this.squares[move.targetSq] = move.promotionPiece;
    }
    if (move.isCastling) {
      const diff = move.targetSq - move.startSq;
      if (diff === 2) { // Kingside
        this.squares[move.targetSq - 1] = this.squares[move.targetSq + 1];
        this.squares[move.targetSq + 1] = '.';
      } else { // Queenside
        this.squares[move.targetSq + 1] = this.squares[move.targetSq - 2];
        this.squares[move.targetSq - 2] = '.';
      }
    }

    // En passant square
    this.enPassant = null;
    if (pieceMoved.toLowerCase() === 'p' && Math.abs(move.targetSq - move.startSq) === 16) {
      this.enPassant = move.startSq + (this.sideToMove === 'b' ? 8 : -8);
    }

    // Castling rights
    if (pieceMoved === 'K') { this.castlingRights.K = false; this.castlingRights.Q = false; }
    if (pieceMoved === 'k') { this.castlingRights.k = false; this.castlingRights.q = false; }
    const fromCorner = ROOK_CORNERS[move.startSq];
    if (fromCorner) this.castlingRights[fromCorner] = false;
    const toCorner = ROOK_CORNERS[move.targetSq];
    if (toCorner) this.castlingRights[toCorner] = false;

    // Half-move clock
    if (pieceMoved.toLowerCase() === 'p' || move.pieceCaptured !== '.') {
      this.halfMoveClock = 0;
    } else {
      this.halfMoveClock++;
    }

    if (this.sideToMove === 'b') this.fullMoveNumber++;
    this.sideToMove = this.sideToMove === 'w' ? 'b' : 'w';
  }

  // Identity of a position for repetition claims: placement, side to move,
  // castling rights and en passant. Deliberately not called from makeMove —
  // the search makes millions of moves and must not pay for this; the game
  // layer calls it once per committed move.
  positionKey() {
    const rights =
      (this.castlingRights.K ? 'K' : '') +
      (this.castlingRights.Q ? 'Q' : '') +
      (this.castlingRights.k ? 'k' : '') +
      (this.castlingRights.q ? 'q' : '') || '-';
    const ep = this._capturableEnPassant();
    return `${this.squares.join('')} ${this.sideToMove} ${rights} ${ep === null ? '-' : ep}`;
  }

  // An en passant square only tells two positions apart when the capture is
  // really on. Otherwise a position reached with a spent en passant square
  // would not match the same position reached without one, and a legitimate
  // repetition would go unnoticed.
  _capturableEnPassant() {
    const ep = this.enPassant;
    if (ep === null) return null;
    const pawn = this.sideToMove === 'w' ? 'P' : 'p';
    // A pawn that could take stands one rank behind the target, on either side.
    const behind = ep + (this.sideToMove === 'w' ? 8 : -8);
    for (const step of [-1, 1]) {
      const from = behind + step;
      if (from < 0 || from > 63) continue;
      if (Math.abs((from % 8) - (ep % 8)) !== 1) continue; // Guards the file wrap
      if (this.squares[from] === pawn) return ep;
    }
    return null;
  }

  unmakeMove(move) {
    this.sideToMove = this.sideToMove === 'w' ? 'b' : 'w';
    if (this.sideToMove === 'b') this.fullMoveNumber--;

    const state = this.history.pop();
    this.castlingRights = state.castlingRights;
    this.enPassant = state.enPassant;
    this.halfMoveClock = state.halfMoveClock;

    // Undo promotion
    if (move.promotionPiece !== '.') {
      this.squares[move.targetSq] = this.sideToMove === 'w' ? 'P' : 'p';
    }

    this.squares[move.startSq] = this.squares[move.targetSq];
    this.squares[move.targetSq] = move.pieceCaptured;

    if (move.isEnPassant) {
      const capSq = move.targetSq + (this.sideToMove === 'w' ? 8 : -8);
      this.squares[capSq] = this.sideToMove === 'w' ? 'p' : 'P';
      this.squares[move.targetSq] = '.';
    }

    if (move.isCastling) {
      const diff = move.targetSq - move.startSq;
      if (diff === 2) {
        this.squares[move.targetSq + 1] = this.squares[move.targetSq - 1];
        this.squares[move.targetSq - 1] = '.';
      } else {
        this.squares[move.targetSq - 2] = this.squares[move.targetSq + 1];
        this.squares[move.targetSq + 1] = '.';
      }
    }
  }
}

// Three occurrences of the same position is a draw. The current position is the
// last key in the list, and the occurrences need not be consecutive.
export function hasThreefoldRepetition(keys) {
  const current = keys[keys.length - 1];
  let seen = 0;
  for (let i = keys.length - 1; i >= 0; i--) {
    if (keys[i] === current && ++seen === 3) return true;
  }
  return false;
}
